docs/strings.md

```diff
@@ -0,0 +1,529 @@
+This is not always a useful abstraction as the definition of a character can, in different contexts, mean:
+
+ * A code unit, the smallest indivisible unit of a single or multi-byte encoding (e.g. `char` or `wchar_t` in C, `unichar` in Objective-C)
+ * A Unicode code point, which may be stored as 1-4 octets, depending on the encoding.
+ * A [grapheme cluster](https://www.unicode.org/reports/tr29/), a sequence of code points that are combined to display something that a human will consider to be a character.
+
+Iteration over strings is therefore very complex.
+Should the unit of iteration be a code unit, a code point, or a grapheme cluster?
+Should iteration canonicalise any of these?
+For example, é can be represented as either a single code point or as an e followed by with an accent combining diacritic, but both representations are regarded as equivalent by unicode.
+
+Mutable and immutable strings
+-----------------------------
+
+A lot of programs depend on immutable strings.
+The Java security model depends on strings being immutable and the `SecurityManager` can be bypassed if a vulnerability allows immutable string objects to be modified.
+Strings are commonly used as keys in collections in most languages (including 'languages' such as POSIX, where they are used to define the filesystem namespace).
+Typically, keys in collections must be immutable for as long as they are being used as keys: modifying a key without notifying the collection will affect the hash or ordering of keys and cause lookup failures.
+
+In Verona, mutability is a property of an instance, not a class.
+This allows a common class for mutable or immutable strings and provides some opportunities for string APIs.
+C's overloading of `const` to mean 'thing that cannot be modified' and 'thing that this function will not modify' causes some problems in the APIs.
+For example, consider the standard library function `strchr`:
+
+```c
+char * strchr(const char *s, int c);
+```
+
+This takes a `const char*` (C string) as an argument and returns a pointer into that string.
+The returned pointer may point to an immutable object but it lacks a `const` qualifier.
+
+In OpenStep, `NSMutableString` is a subclass of (immutable) `NSString`, which allows APIs to be implemented as methods on `NSString` or as functions that take `NSString` arguments.
+This can lead to unsafe down-casts and the fact that a mutable string is a kind of immutable string causes bugs when mutable strings are accidentally used as immutable strings and modified unexpectedly.
+
+In Verona, we can differentiate between `imm` parameters, where the callee may capture the argument and requires that it is never mutated, and `readonly` parameters, where the string may be mutable but the caller promises not to modify them and the compiler enforces that guarantee.
+Collections using strings as keys should always take an `imm & String`, whereas functions that query properties of a string should take a `readonly & String`.
```

Comment:
I'm still unsure of the difference between a `readonly` variable and an `immutable` view.

They're both a reference to a variable where you can't change the value of that variable through this reference.

I think this is a discussion for another thread, but perhaps this section is a bit loose in that it sounds important to consider but is not really linked with the surrounding sections...

Reply:
I tried factoring out a `StringView` from `String`, but I think a string view needs basically all of the methods of a string.  You can't have an `imm & String` (or `StringView`) that refers to a range in a larger string.  You can have a `readonly & String` that has storage that aliases another string.